### DISCUSSION TITLE
Stats for time writing emails

### DISCUSSION THREAD

**Discussion by webber-naut** [devmount/third-stats#408] [Q&A] [answered]
I didn't see anything when I searched for an answer in this fourm. 
And if this already exists in the plugin, forgive me I don't actually have Thunderbird installed or the add on installed. I plan on transitioning shortly.

Is it possible to track time spent writing emails and that data would be tied to specific emails? This would include the time you hit reply to the time you hit send. Or if you compose a new email once the email is entered into the TO field it would start the timer up to the point of hitting send? 

I would like to keep track of time spent writing emails to specific email addresses (or all email addresses is fine too)

**Comment by devmount** [comment, marked as answer]
No worries, this hasn't been asked before. Thank you for this great idea! Unfortunately that's something ThirdStats can't do right now and won't in the near future. ThirdStats always only analyses the current state, which are all messages that are available at the time of processing. It doesn't do measurements over time or holds legacy data (like e.g. deleted messages).

It's a good feature for the long term todo list though, so thanks for the request.

**Reply by webber-naut** [reply]
Thanks for the quick reply. Maybe this feature will make it's way into your add on at some point. In the mean time I might take a go at this myself, I have some Firefox extension experience, I imagine it's very similar process/code base. I do think it would be a great additional feature and very useful to everyone using your add on.

**Reply by devmount** [reply]
Go for it! And let us know if you have anything ready to try out 👍🏻

**Reply by nozebacle** [reply]
I've got zero experience developing Thunderbird or Firefox extensions, but I wonder if the same mechanism that is used to join emails into conversations, could be used to measure something similar to what @webber-naut  is proposing? In my case, I'd love to find out how fast or how slow I respond to messages and maybe do some analysis on that data (are messages coming from certain accounts processed faster than others? what is the best day to write me to get a quick answer? emails in which folders have the fastest response times?) and also see how many emails didn't require a response.

**Reply by devmount** [reply]
@nozebacle Wow, analyzing email threads/conversations is a great idea! If it's possible to retrieve thread data from web extension APIs it shouldn't be too difficult to calculate the average time between answers and map it to the corresponding contact or folder 😍 Unfortunately I couldn't find any API about thread data, so I [posted in Thunderbirds topicbox](https://thunderbird.topicbox.com/groups/addons/T679436983ba3445c/get-thread-data-for-messages).

@webber-naut Would this be usable for you too or are you only interested in the time needed to actually write a message?

**Reply by webber-naut** [reply]
The idea of getting specific email stats independently would be super useful, which is what I think @nozebacle is proposing.

But my use case is tracking how much time is spent writing emails from both compose, and reply to all or specific email addresses. (thinking a whitelist option in the settings somewhere would be the way to go, or a default track all checkbox and whitelist option for over ride)

The documentation for Add On's seems to be very limited, I guess I have to dig into the actual source code on github. ChatGPT didn't seem very helpful either as I think it has very outdated knowledge on the topic. Playing around a little with lots of unsuccessful attempts, I managed to grab the email address with messenger.compose.onComposeStateChanged.addListener()  but it only works with reply. When using compose aka "write", the email address comes back blank. I think I might have to write a whole experiment API function. Javascript isn't my favorite code base.

If you have any links, or documentation references for compose I would much appreciate it.
This is my reference I was using: https://webextension-api.thunderbird.net/en/102/compose.html

**Reply by devmount** [reply]
I see, yes that's of course two different kind of stats which should be handled separately.

Your reference is the official one I'm using too. In addition you could take a look into https://github.com/thundernest/sample-extensions, there are a couple of examples for using the compose API too. Also you always have the possibility to ask questions in [Thunderbirds TopicBox for add-on devs](https://thunderbird.topicbox.com/groups/addons).
